src/relay/frontend/graph_runtime_codegen.h

```diff
@@ -0,0 +1,607 @@
+
+#include "../backend/compile_engine.h"
+#include "utils.h"
+
+namespace tvm {
+namespace relay {
+namespace frontend {
+
+class _Node;
+class _InputNode;
+class _OpNode;
+
+using IntegerArray = Array<Integer>;
+using ShapeVector = std::vector<std::vector<int64_t>>;
+using Attrs = std::unordered_map<std::string, dmlc::any>;
+using _NodePtr = std::shared_ptr<_Node>;
+using _InputNodePtr = std::shared_ptr<_InputNode>;
+using _OpNodePtr = std::shared_ptr<_OpNode>;
+
+/*! \brief Lowered outputs */
+struct LoweredOutput {
+  std::string graph_json;
+  std::unordered_map<std::string, std::vector<LoweredFunc>> lowered_funcs;
+  std::unordered_map<std::string, tvm::runtime::NDArray> params;
+};
+
+/*! \brief Node types */
+enum _NodeType {
+  kNop_,
+  kInputNode_,
+  kOpNode_,
+};
+
+/*! \brief Node Reference */
+class _NodeRef {
+ public:
+  _NodeRef() {}
+
+  _NodeRef(int ident, int index, int version = 0)
+      : ident_(ident), index_(index), version_(version) {}
+
+  inline void Save(dmlc::JSONWriter* writer) const {
+    writer->BeginArray();
+    writer->WriteArrayItem(ident_);
+    writer->WriteArrayItem(index_);
+    writer->WriteArrayItem(version_);
+    writer->EndArray();
+  }
+
+  inline void Load(dmlc::JSONReader* reader) {}
+
+ protected:
+  int ident_;
+  int index_{0};
+  int version_{0};
+};
+
+/*! \brief Base Node class */
+class _Node {
+ public:
+  _Node() {}
+  virtual void Save(dmlc::JSONWriter* writer) const {}
+  virtual void Load(dmlc::JSONReader* reader) {}
+  virtual _NodeType Type() const { return kNop_; }
+  virtual ~_Node() {}
+
+ public:
+  int num_outputs_{1};
+  std::string name_;
+  std::unordered_map<std::string, dmlc::any> attrs_;
+};
+
+/*! \brief Input Node */
+class _InputNode : public _Node {
+ public:
+  _InputNode() {}
+  _InputNode(const std::string& name, const Attrs& attrs) {
+    name_ = name;
+    attrs_ = attrs;
+  }
+
+  _NodeType Type() const override { return kInputNode_; }
+  void Load(dmlc::JSONReader* reader) override {}
+  void Save(dmlc::JSONWriter* writer) const override {
+    const std::string op_name{"null"};
+    writer->BeginObject();
+    writer->WriteObjectKeyValue("op", op_name);
+    writer->WriteObjectKeyValue("name", this->name_);
+    writer->WriteObjectKeyValue("inputs", std::list<int>());
+    writer->EndObject();
+  }
+  static std::shared_ptr<_Node> make_node_ptr(const std::string& name, const Attrs& attrs) {
+    auto ptr = std::make_shared<_InputNode>(name, attrs);
+    return std::dynamic_pointer_cast<_Node>(ptr);
+  }
+};
+
+/*! \brief Op Node */
+class _OpNode : public _Node {
+ public:
+  _OpNode() {}
+  _OpNode(const std::string& name, const Attrs& nd_attrs, const std::string& op_name,
+          const std::vector<_NodeRef>& inputs, const Attrs& attrs, size_t num_outputs = 1) {
+    name_ = name;
+    attrs_ = nd_attrs;
+    op_name_ = op_name;
+    inputs_ = inputs;
+    op_attrs_ = attrs_;
+    num_outputs_ = num_outputs;
+    op_attrs_["func_name"] = op_name_;
```

Comment:
You need to move this down to to where you actually need to `Save` to JSON, because some of the properties will get changed (eg. `num_outputs`). 

Here is a patch that you can apply
```
diff --git a/src/relay/frontend/graph_runtime_codegen.h b/src/relay/frontend/graph_runtime_codegen.h
--- a/src/relay/frontend/graph_runtime_codegen.h
+++ b/src/relay/frontend/graph_runtime_codegen.h
@@ -116,10 +116,6 @@
     inputs_ = inputs;
     op_attrs_ = attrs_;
     num_outputs_ = num_outputs;
-    op_attrs_["func_name"] = op_name_;
-    op_attrs_["flatten_data"] = std::string("0");
-    op_attrs_["num_inputs"] = std::to_string(inputs_.size());
-    op_attrs_["num_outputs"] = std::to_string(num_outputs_);
   }
   std::string op_name_;
   std::vector<_NodeRef> inputs_;
@@ -130,10 +126,15 @@
   }
   void Load(dmlc::JSONReader *reader) override {}
   void Save(dmlc::JSONWriter *writer) const override {
+    Attrs attrs = op_attrs_;
+    attrs["func_name"] = this->op_name_;
+    attrs["flatten_data"] = std::string("0");
+    attrs["num_inputs"] = std::to_string(this->inputs_.size());
+    attrs["num_outputs"] = std::to_string(this->num_outputs_);
     writer->BeginObject();
     writer->WriteObjectKeyValue("op", op_type_name_);
     writer->WriteObjectKeyValue("name", name_);
-    writer->WriteObjectKeyValue("attrs", op_attrs_);
+    writer->WriteObjectKeyValue("attrs", attrs);
     writer->WriteObjectKeyValue("inputs", this->inputs_);
     writer->EndObject();
   }
```

Here is an example that you can test. Without this patch, json will have the 1 instead 2 output number. 
```
fn (%X0: Tensor[(7, 8), float32],
    %X1: Tensor[(7, 8), float32],
    %X2: Tensor[(7, 8), float32]) {
  %0 = (%X0, %X1, %X2)
  %1 = concatenate(%0, axis=1)
  %2 = (%X0, %X1, %X2)
  %3 = concatenate(%2, axis=1)
  %4 = reshape(%3, newshape=[0, 3, -1])
  %5 = (%1, %4)
  %5
}
```